presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/api/PackageAPI.java

```diff
@@ -36,87 +35,100 @@
-            return Response.status(BAD_REQUEST).entity(text("Invalid package type for presto-server. Package must be .tar.gz or .rpm")).build();
+        if ("".equals(locationToFetchPackage)) {
+            LOGGER.error("Url is empty or null");
+            return Response.status(BAD_REQUEST).entity("Expected URL in the body").build();
         }
         try {
+            if (isInstalled()) {
+                LOGGER.error("Presto is already installed.");
+                return Response.status(CONFLICT).entity("Presto is already installed.").build();
+            }
             URL url = new URL(locationToFetchPackage);
-            new Thread(new PrestoAsynchronousCommand(new PrestoInstaller(installedPackageType, url, checkDependencies))).start();
+            new Thread(new PrestoAsynchronousCommand(new PrestoInstaller(RPM, url, checkDependencies))).start();
+        }
+        catch (PrestoManagerException e) {
+            LOGGER.error("Failed to ascertain whether presto is running", e);
+            return Response.status(INTERNAL_SERVER_ERROR).entity("Failed to ascertain whether presto is already installed.").build();
         }
         catch (MalformedURLException e) {
             LOGGER.error("Invalid url: {}", locationToFetchPackage, e);
-            return Response.status(BAD_REQUEST).entity(text("Invalid url")).build();
+            return Response.status(BAD_REQUEST).entity("Invalid url").build();
         }
-        return Response.status(ACCEPTED).entity(text("Check status to make sure that presto is installed")).build();
+        return Response.status(ACCEPTED).entity("Presto is being installed. " +
+                "Please check back later using status api to make sure that install is successful.").build();
     }
 
     @POST
-    @Consumes({MediaType.TEXT_PLAIN})
-    @Produces({MediaType.TEXT_PLAIN})
+    @Consumes({TEXT_PLAIN})
+    @Produces({TEXT_PLAIN})
     @ApiOperation(value = "Upgrade presto")
     @ApiResponses(value = {
             @ApiResponse(code = 202, message = "Acknowledged request"),
-            @ApiResponse(code = 400, message = "Bad package format or Invalid url")
+            @ApiResponse(code = 400, message = "Invalid url"),
+            @ApiResponse(code = 409, message = "Presto is running")
     })
-    public synchronized Response upgrade(@NotNull @ApiParam("Url to fetch package") String locationToFetchPackage,
-            @QueryParam("checkDependencies") @DefaultValue("true") @ApiParam("If false, disables dependency checking") boolean checkDependencies)
+    public synchronized Response upgrade(@ApiParam("Url to fetch package") String locationToFetchPackage,
+            @QueryParam("checkDependencies") @DefaultValue("true") @ApiParam("If false, disables dependency checking") boolean checkDependencies,
+            @QueryParam("retainConfigs") @DefaultValue("true") @ApiParam("If false, config files are not retained") boolean retainConfigs)
```

Comment:
I would say "preserve" was better than "retain" ("save" or "keep" would also work well). This only matters for the value in the `QueryParam` annotation.
Actually, reversing the boolean might be a better idea: "resetConfig" is about as clear and the others, and defaults to `false`, which is the default value for `boolean`s, which makes it easier for the user to guess the default, and more importantly, should save the user from typos (assuming that it uses `Boolean#parseBoolean(String)` to convert the parameter to a string).

And please remove the "s", we're only dealing with one configuration.